File: actions_py/actions_py/move_robot_server.py
#!/usr/bin/env python3
import rclpy
import time
import threading
from rclpy.node import Node
from rclpy.action import ActionServer, GoalResponse, CancelResponse
from rclpy.action.server import ServerGoalHandle
from my_robot_interfaces.action import MoveRobot
from rclpy.executors import MultiThreadedExecutor
from rclpy.callback_groups import ReentrantCallbackGroup


class MoveRobotServerNode(Node): # MODIFY NAME
    def __init__(self):
        super().__init__("move_robot_server") # MODIFY NAME
        self.robot_position_ = 50
        self.goal_handle_: ServerGoalHandle = None
        self.goal_lock_ = threading.Lock()
        self.move_robot_server_ = ActionServer(
            self,
            MoveRobot,
            "move_robot",
            goal_callback=self.goal_callback,
            cancel_callback=self.cancel_callback,
            execute_callback=self.execute_callback)
        self.get_logger().info("Action server has been started")
        self.get_logger().info(f"Robot position: {self.robot_position_}")
    
    def goal_callback(self, goal_request: MoveRobot.Goal):
        self.get_logger().info("Received a goal")
        
        # Validate the goal request
        if not (0 <= goal_request.position <= 100 and 0 <= goal_request.velocity <= 100):
            self.get_logger().info("Rejecting the goal")
            return GoalResponse.REJECT
        
        #Policy prempt existing goal when receiving new goal
        with self.goal_lock_:
            if self.goal_handle_ is not None and self.goal_handle_.is_active:
                self.get_logger().info("Abort current goal and accept new goal")
                self.goal_handle_.abort()
        
        self.get_logger().info("Accepting a goal")
        return GoalResponse.ACCEPT
    
    def cancel_callback(self, goal_handle: ServerGoalHandle):
        self.get_logger().info("Received a cancel request")
        return CancelResponse.ACCEPT # or REJECT

    def execute_callback(self, goal_handle: ServerGoalHandle):
        with self.goal_lock_:
            self.goal_handle_ = goal_handle    
        
        # Get request from goal
        goal_position = goal_handle.request.position
        velocity = goal_handle.request.velocity
        
        # Execute the action
        self.get_logger().info("Executing the goal")
        feedback = MoveRobot.Feedback()
        result = MoveRobot.Result()
        while goal_position != self.robot_position_:
            if not goal_handle.is_active:
                result.position = self.robot_position_
                return result
            
            if goal_handle.is_cancel_requested:
                self.get_logger().info("Canceling the goal")
                goal_handle.canceled()
                result.position = self.robot_position_
                result.message = "Robot movement cancelled"
                return result
                
            if goal_position > self.robot_position_:
                if (goal_position - self.robot_position_) > velocity:
                    self.robot_position_ += velocity
                    self.get_logger().info(f"{self.robot_position_}")
                else:
                    self.robot_position_ = goal_position
                    self.get_logger().info(f"{self.robot_position_}")
            elif goal_position < self.robot_position_:
                if (self.robot_position_ - goal_position) > velocity:
                    self.robot_position_ -= velocity
                    self.get_logger().info(f"{self.robot_position_}")
                else:
                    self.robot_position_ = goal_position
                    self.get_logger().info(f"{self.robot_position_}")
            feedback.current_position = self.robot_position_
            goal_handle.publish_feedback(feedback)
            time.sleep(1.0)
        
        # Once done, set the goal final state
        goal_handle.succeed()
        
        # and send the result
        result.position = self.robot_position_
        result.message = "Robot has arrived at destination"
        return result

def main(args=None):
    rclpy.init(args=args)
    node = MoveRobotServerNode() # MODIFY NAME
    rclpy.spin(node, MultiThreadedExecutor())
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: actions_cpp/src/move_robot_server.cpp
#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"
#include "my_robot_interfaces/action/move_robot.hpp"

using MoveRobot = my_robot_interfaces::action::MoveRobot;
using MoveRobotGoalHandle = rclcpp_action::ServerGoalHandle<MoveRobot>;
using namespace std::placeholders;

class MoveRobotServerNode : public rclcpp::Node {
public:
    MoveRobotServerNode() : Node("move_robot_server") {
        robot_position_ = 50;
        cb_group_ = this->create_callback_group(rclcpp::CallbackGroupType::Reentrant);
        move_robot_server_ = rclcpp_action::create_server<MoveRobot>(
            this,
            "move_robot",
            std::bind(&MoveRobotServerNode::goal_callback, this, _1, _2),
            std::bind(&MoveRobotServerNode::cancel_callback, this, _1),
            std::bind(&MoveRobotServerNode::handle_accepted_callback, this, _1),
            rcl_action_server_get_default_options(),
            cb_group_
        );
    RCLCPP_INFO(this->get_logger(), "Action server has been started");
    }

private:

    rclcpp_action::GoalResponse goal_callback(
        const rclcpp_action::GoalUUID &uuid,
        std::shared_ptr<const MoveRobot::Goal> goal) {
        
        (void)uuid;
        (void)goal;

        // Policy: Abort current goal if new goal is received
        { std::lock_guard<std::mutex> lock(mutex_);
            if (goal_handle_) {
                if (goal_handle_->is_active()) {
                    RCLCPP_WARN(this->get_logger(), "New goal accepted, existing goal aborted");
                    preempted_goal_id_ = goal_handle_->get_goal_id();
                }
            }
        }
        return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
    }
    rclcpp_action::CancelResponse cancel_callback(
        const std::shared_ptr<MoveRobotGoalHandle> goal_handle) {
        
        (void)goal_handle;
        return rclcpp_action::CancelResponse::ACCEPT;
    }

    void handle_accepted_callback(const std::shared_ptr<MoveRobotGoalHandle> goal_handle) {
        execute(goal_handle);
    }

    void execute(const std::shared_ptr<MoveRobotGoalHandle> goal_handle) {
        {
            std::lock_guard<std::mutex> lock(mutex_);
            this->goal_handle_ = goal_handle;
        }
        
        int goal_position = goal_handle->get_goal()->position;
        int velocity = goal_handle->get_goal()->velocity;
        auto feedback = std::make_shared<MoveRobot::Feedback>();
        auto result = std::make_shared<MoveRobot::Result>();
        rclcpp::Rate loop_rate(1.0);

        while (rclcpp::ok()) {
            int difference = goal_position - robot_position_;
            {
                std::lock_guard<std::mutex> lock(mutex_);
                if (goal_handle->get_goal_id() == preempted_goal_id_) {
                    result->position = robot_position_;
                    goal_handle->abort(result);
                    return;
                }
            }

            if (goal_handle->is_canceling()) {
                result->position = robot_position_;
                result->message = "Goal cancelled";
                goal_handle->canceled(result);
                return;
            }

            if (!difference) {
                RCLCPP_INFO(this->get_logger(), "Robot has reach it's goal of: %d", goal_position);
                result->position = goal_position;
                goal_handle->succeed(result);
                return;
            }
            else if (difference > 0) {
                robot_position_ += std::min(difference, velocity);
            }
            else {
                robot_position_ += std::max(difference, -velocity);
            }
            RCLCPP_INFO(this->get_logger(), "Robot is at: %d", robot_position_);
            feedback->current_position = robot_position_;
            goal_handle->publish_feedback(feedback);

            loop_rate.sleep();
        }
        
    }

    rclcpp_action::Server<MoveRobot>::SharedPtr move_robot_server_;
    std::shared_ptr<MoveRobotGoalHandle> goal_handle_;
    rclcpp_action::GoalUUID preempted_goal_id_;
    rclcpp::CallbackGroup::SharedPtr cb_group_;
    std::mutex mutex_;
    int robot_position_;
};

int main(int argc, char **argv) {
    rclcpp::init(argc, argv);
    auto node = std::make_shared<MoveRobotServerNode>();
    rclcpp::executors::MultiThreadedExecutor executor;
    executor.add_node(node);
    executor.spin();
    rclcpp::shutdown();
    return 0;
}